crt0stack/src/lib.rs

```diff
@@ -0,0 +1,486 @@
+                self.save(0u8)?;
+                (BasePlatform, self.save_slice(&x.as_bytes())? as _)
+            }
+            Entry::ExecFilename(x) => {
+                self.save(0u8)?;
+                (ExecFilename, self.save_slice(&x.as_bytes())? as _)
+            }
+            Entry::Random(x) => (Random, self.save_slice(&x)? as _),
+            Entry::ExecFd(v) => (ExecFd, v as _),
+            Entry::PHdr(v) => (PHdr, v as _),
+            Entry::PHent(v) => (PHent, v as _),
+            Entry::PHnum(v) => (PHnum, v as _),
+            Entry::Pagesize(v) => (Pagesize, v as _),
+            Entry::Base(v) => (Base, v as _),
+            Entry::Flags(v) => (Flags, v as _),
+            Entry::Entry(v) => (Entry, v as _),
+            Entry::Notelf(v) => (NotElf, if v { 1 } else { 0 }),
+            Entry::Uid(v) => (Uid, v as _),
+            Entry::EUid(v) => (EUid, v as _),
+            Entry::Gid(v) => (Gid, v as _),
+            Entry::EGid(v) => (EGid, v as _),
+            Entry::HWCap(v) => (HWCap, v as _),
+            Entry::ClockTick(v) => (ClockTick, v as _),
+            Entry::Secure(v) => (Secure, if v { 1 } else { 0 }),
+            Entry::HWCap2(v) => (HWCap2, v as _),
+        };
+        self.push_item(key)?;
+        self.push_item(value)?;
+        Ok(())
+    }
+
+    pub fn done(mut self) -> Result<Handle<'a>> {
+        self.push_item(Key::default())?;
+        self.push_item(Type::default())?;
+
+        let start_ptr = {
+            // at the end, copy the items of the data section from the bottom to the top of the stack
+            let (bottom, top) = self.stack.split_at_mut(self.items);
+
+            let (_prefix, src, _suffix) = unsafe { bottom.align_to_mut::<Type>() };
+
+            let (_prefix, dst, _suffix) = {
+                let end = self.data - self.items;
+                unsafe { top[0..end].align_to_mut::<Type>() }
+            };
+            let start = dst.len() - src.len();
+            let end = dst.len();
+            dst[start..end].copy_from_slice(src);
+            &dst[start] as *const usize as *const ()
+        };
+        Ok(Handle(self.stack, start_ptr))
+    }
+}
+
+impl<'a> Builder<'a, Env> {
+    // Saves the env string to the data section and creates
+    // a reference to it in the items section.
+    pub fn push(&mut self, env: &str) -> Result<()> {
+        self.save(0_u8)?; // c-str zero byte
+        let p = self.save_slice(env.as_bytes())?;
+        self.push_item(p)
+    }
+
+    pub fn next(mut self) -> Result<Builder<'a, Aux>> {
+        // last environ is NULL
+        self.push_item(0_usize)?;
+        Ok(Builder {
+            stack: self.stack,
+            data: self.data,
+            items: self.items,
+            state: PhantomData,
+        })
+    }
+}
+
+impl<'a> Builder<'a, Arg> {
+    pub fn new(stack: &'a mut [u8]) -> Self {
+        let len = stack.len();
+        Self {
+            stack,
+            data: len,
```

Comment:
```suggestion
            data: stack.len(),
```

Reply:
error[E0502]: cannot borrow `*stack` as immutable because it is also borrowed as mutable
   --> src/lib.rs:123:19
    |
118 |   impl<'a> Builder<'a, Arg> {
    |        -- lifetime `'a` defined here
...
121 | /         Self {
122 | |             stack,
    | |             ----- mutable borrow occurs here
123 | |             data: stack.len(),
    | |                   ^^^^^ immutable borrow occurs here
124 | |             items: core::mem::size_of::<usize>(),
125 | |             state: PhantomData,
126 | |         }
    | |_________- returning this value requires that `*stack` is borrowed for `'a`